### PR TITLE
feat: "edit workspace"

### PR DESCRIPTION
**Pull request by bctnry**

### CODE REVIEW

**Comment by yjqww6** [comment]
I see workspaces are referenced by indices rather than their names in current implementation. Workspaces with duplicated names seem ok currently, but I am not sure whether it is by design. @Syntacticlosure 

PS: I prefer not to duplicate the implementation with "new workspace".

**Comment by Syntacticlosure** [comment]
LGTM

File: private/popup-menu.rkt [hunk 1]
```diff
@@ -84,33 +85,55 @@
...
+                    [callback (λ (list-box event)
+                                (when (equal? (send event get-event-type) 'list-box-dclick)
+                                  (begin
+                                    ;; NOTE: this will only take the first selected workspace if
+                                    ;; multiple workspace is selected, as occasions where you need
+                                    ;; to edit multiple workspace at once is kinda rare.
+                                    ;; `get-string-selection` only gets the first column.
+                                    ;; `get-data` with `get-selection` somehow does not work (will
+                                    ;; only returns #f)
+                                    (let ((item (send list-box get-string-selection)))
```
Comment (Syntacticlosure) [inline private/popup-menu.rkt]: use racket coding style: (let ((x v)) -> (let ([x v])

File: private/popup-menu.rkt [hunk 2]
```diff
@@ -84,33 +85,55 @@
...
+                                                                           (preferences:get 'files-viewer:workspaces))]
+                                                                         #:unless (= (first s) i))
+                                                                x))
                                              (refresh)
                                              ))]))
+    (define edit-button (new button% [parent bp]
+                             [label "Edit workspace"]
+                             [callback (λ (c e)
+                                         (begin
+                                           (let ((item (send ws get-string-selection)))
```
Comment (Syntacticlosure) [inline private/popup-menu.rkt]: same as above

File: private/popup-menu.rkt [hunk 3]
```diff
@@ -84,33 +85,55 @@
...
               (map second prefs))))
+      
     (define ws (new list-box%
                     [parent this][choices '()]
                     [style '(column-headers single)]
-                    [label ""][columns '("Name" "Path")]))
+                    [label ""][columns '("Name" "Path")]
+                    [callback (λ (list-box event)
+                                (when (equal? (send event get-event-type) 'list-box-dclick)
+                                  (begin
```
Comment (sorawee) [inline private/popup-menu.rkt]: No need for `begin`.

File: private/popup-menu.rkt [hunk 4]
```diff
@@ -84,33 +85,55 @@
...
+                                                                          (in-indexed
+                                                                           (preferences:get 'files-viewer:workspaces))]
+                                                                         #:unless (= (first s) i))
+                                                                x))
                                              (refresh)
                                              ))]))
+    (define edit-button (new button% [parent bp]
+                             [label "Edit workspace"]
+                             [callback (λ (c e)
+                                         (begin
```
Comment (sorawee) [inline private/popup-menu.rkt]: No need for `begin`.

File: private/utils.rkt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#lang racket
+;;; aux funcs.
+(provide (all-defined-out))
+
+;; NOTE: assumes `m` is a (Listof (List KeyType ValueType)).
+;; this func will not set! for you; you have to do it manually.
+(define (assocmap-set m k v)
+  (cond
+    ((null? m) m)
```
Comment (sorawee) [inline private/utils.rkt]: use racket coding style: `(cond (c e) ...)` -> `(cond [c e] ...)`

Comment (sorawee) [reply private/utils.rkt]: Though really, you can just write 

```
(dict-set alist k (list v))
```

instead. No need to define a new function.

**Comment by Syntacticlosure** [comment]
> I see workspaces are referenced by indices rather than their names in current implementation. Workspaces with duplicated names seem ok currently, but I am not sure whether it is by design. @Syntacticlosure
> 
> PS: I prefer not to duplicate the implementation with "new workspace".

You are right. Though current implementation permits duplicated names of workspaces, it is sensible to forbid creating workspaces with same names in the future versions.

**Pull request merged**